test/fixtures/rules/std_rule_cases/L044.yml

```diff
@@ -327,3 +327,46 @@ test_fail_unknown_number_of_result_columns_18:
         where not coalesce(_fivetran_deleted, false)
     )
     select * from final
+
+test_pass_nested_ctes_1:
+  # Test for issue 1984
+  pass_str: |
+    with a as (
+        with b as (
+            select 1 from c
+        )
+
+        select * from b
+    )
+
+    select * from a
```

Comment:
for completeness, maybe add extra test cases:

```sql
with a as (
    with b as (
        select * from c
    )

    select 1 from b
)

select * from a
```

```sql
with a as (
    with b as (
        select * from c
    )

    select * from b
)

select 1 from a
```

Reply:
don't think we need a test for it, but also checked 3 levels of nested with statements and it works perfectly 👍

Reply:
I'll add those. 👍🏽

Reply:
bit of a weird one but worth looking at if you get the chance, the rule seems to get tripped up if you mix nesting and same level ctes:
![image](https://user-images.githubusercontent.com/80432516/143880775-eef11aee-a28c-492b-a4a2-b35b5aac1766.png)

Reply:
Another really weird edge case that also trips up L045:
![image](https://user-images.githubusercontent.com/80432516/143882373-3f0f8553-0c25-4ff2-945a-e21d2d2b1365.png)

Reply:
Good test cases! I'm not sure what's happening in the first one. In the last one, I think L044 does not "see" CTEs defined at an outer scope, only the same level. I'll look into that and add those two cases.

I'll also create an issue about updating L045. Its design does not consider nested CTEs, either. In working on that other rule, I may go back and rework parts of L044 as well, if I see benefits to doing that.

Reply:
Created this issue: #1999

Reply:
perfect thanks 😄

Reply:
After taking a bit closer look at the two new test cases, I think I'll go ahead and rework `SelectCrawler` now. Thus far, I had added some of the nested CTE logic directly in L044, but it has bugs, and if I've got to fix it anyway, I may as well improve the design as well. 👍🏽

Reply:
makes sense!

had a couple more ideas for unit tests for L044 that we don't have in L044.yml yet, both of these pass currently but worth adding them 😄 

```sql
with a as (
    select 1 from b
)

select * from (
    select * from a
);
```

```sql
with b as (
    select 1 from c
)

select * from (
    with a as (
        select * from b
    )

    select * from a
);
```